node/src/components/contract_runtime/rewards.rs

```diff
@@ -0,0 +1,380 @@
+use std::{collections::BTreeMap, ops::Range, sync::Arc};
+
+use either::Either;
+use futures::stream::{self, StreamExt as _, TryStreamExt as _};
+use num_rational::Ratio;
+use num_traits::Zero;
+
+use crate::{
+    contract_runtime::EraValidatorsRequest,
+    effect::{
+        requests::{ContractRuntimeRequest, StorageRequest},
+        EffectBuilder,
+    },
+    types::{Block, BlockHash, Chainspec},
+};
+use casper_execution_engine::core::engine_state::{self, GetEraValidatorsError};
+use casper_hashing::Digest;
+use casper_types::{EraId, ProtocolVersion, PublicKey, U512};
+
+pub(crate) trait ReactorEventT:
+    Send + From<StorageRequest> + From<ContractRuntimeRequest>
+{
+}
+
+impl<T> ReactorEventT for T where T: Send + From<StorageRequest> + From<ContractRuntimeRequest> {}
+
+struct ErasInfo(BTreeMap<EraId, EraInfo>);
+
+/// The era information needed in the rewards computation:
+struct EraInfo {
+    weights: BTreeMap<PublicKey, U512>,
+    total_weights: U512,
+    reward_per_round: Ratio<U512>,
+}
+
+#[derive(Debug)]
+pub enum RewardsError {
+    /// There were an error while trying to get the eras information.
+    PopulateError(PopulateError),
+    /// We got a block height which is not in the era range it should be in (should not happen).
+    HeightNotInEraRange(u64),
+    /// The era is not in the range we have (should not happen).
+    EraIdNotInEraRange(EraId),
+    /// The validator public key is not in the era it should be in (should not happen).
+    ValidatorKeyNotInEra(Box<PublicKey>),
+    /// We didn't have a required switch block.
+    MissingSwitchBlock(EraId),
+}
+
+/// We could not fetch the asked information for the given eras.
+#[derive(Debug)]
+pub enum PopulateError {
+    FailedToFetchBlock(BlockHash),
+    FailedToFetchEra(GetEraValidatorsError),
+    /// Fetching the era validators succedeed, but no info is present (should not happen).
+    /// The `Digest` is the one that was queried.
+    NoEraReturned(Digest),
+    FailedToFetchTotalSupply(engine_state::Error),
+    FailedToFetchSeigniorageRate(engine_state::Error),
+}
+
+impl ErasInfo {
+    /// `block_hashs` is an iterator over the era ID to get the information about + the block
+    /// hash to query to have such information (which may not be from the same era).
+    pub async fn populate<REv: ReactorEventT>(
+        effect_builder: EffectBuilder<REv>,
+        protocol_version: ProtocolVersion,
+        block_hashs: impl IntoIterator<Item = (EraId, BlockHash)>,
```

Comment:
I think we generally pluralize these as "hashes", changing that would make it easier to find in search

Reply:
It's even "hashes" later in this file

Reply:
Done